ambry-server/src/main/java/com/github/ambry/server/StatsManager.java

```diff
@@ -435,53 +478,108 @@ long getPermanentDeleteTombstoneSize() {
+          while (!cancelled && iterator.hasNext()) {
+            PartitionId partitionId = iterator.next();
+            logger.info("Aggregating account stats for local report started for store {}", partitionId);
+            collectAndAggregateAccountStats(aggregatedSnapshot, partitionId, unreachablePartitions);
+          }
+          aggregatedSnapshot.updateValue();
+          List<String> unreachableStores = examineUnreachablePartitions(unreachablePartitions);
+          updateAggregatedDeleteTombstoneStats();
+          metrics.initDeleteStatsGaugesIfNeeded();
+          if (!cancelled) {
+            metrics.totalFetchAndAggregateTimeMs.update(time.milliseconds() - totalFetchAndAggregateStartTimeMs);
+            StatsHeader statsHeader =
+                new StatsHeader(StatsHeader.StatsDescription.STORED_DATA_SIZE, time.milliseconds(),
+                    partitionToReplicaMap.keySet().size(),
+                    partitionToReplicaMap.keySet().size() - unreachableStores.size(), unreachableStores);
+            // First write to account stats
+            StatsWrapper statsWrapper = new StatsWrapper(statsHeader, aggregatedSnapshot);
+            if (accountStatsMySqlStore != null) {
+              accountStatsMySqlStore.storeAccountStats(statsWrapper);
+            }
+            publish(statsWrapper);
+            logger.info("Local account stats snapshot published to {}", statsOutputFile.getAbsolutePath());
+          }
+        } catch (Exception | Error e) {
+          metrics.statsAggregationFailureCount.inc();
+          logger.error("Exception while aggregating account stats for local report. Stats output file path - {}",
+              statsOutputFile.getAbsolutePath(), e);
+        } finally {
           if (accountStatsMySqlStore != null) {
-            accountStatsMySqlStore.storeAccountStats(statsWrapper);
+            accountStatsMySqlStore.closeConnection();
           }
-          publish(statsWrapper);
-          logger.info("Local stats snapshot published to {}", statsOutputFile.getAbsolutePath());
         }
-      } catch (Exception | Error e) {
-        metrics.statsAggregationFailureCount.inc();
-        logger.error("Exception while aggregating stats for local report. Stats output file path - {}",
-            statsOutputFile.getAbsolutePath(), e);
-      } finally {
-        if (accountStatsMySqlStore != null) {
+      }
+    }
+
+    void cancel() {
+      cancelled = true;
+    }
+  }
+
+  /**
+   * Runnable class that collects, publishes partition class stats to mysql database.
+   */
+  class PartitionClassStatsPublisher implements Runnable {
+    private volatile boolean cancelled = false;
+    private final AccountStatsMySqlStore accountStatsMySqlStore;
+
+    PartitionClassStatsPublisher(AccountStatsMySqlStore accountStatsMySqlStore) {
+      this.accountStatsMySqlStore = accountStatsMySqlStore;
+    }
+
+    @Override
+    public void run() {
+      synchronized (accountStatsMySqlStore) {
```

Comment:
Why can we assume that accountStatsMySqlStore is non-null here but not in the other task?
Also, do the actions of PartitionClassStatsPublisher and AccountStatsPublisher need to be serialized with each other (using same lock)?

Reply:
This is guaranteed, since we have to enable mysql report before we can enable the partition class reports. We have a check in StatsManagerConfig.java.